src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1060,6 +1063,21 @@ CreateDistributedPlan(uint64 planId, Query *originalQuery, Query *query, ParamLi
 	 */
 	if (list_length(subPlanList) > 0 || hasCtes)
 	{
+		/*
+		 * recursive planner should handle all the tree from bottom to
+		 * top at single pass. i.e. It should have already recursively planned all
+		 * required parts in its first pass. Hence, we expect previousSubPlanList
+		 * to be empty. Otherwise, this means we have bug at recursive planner,
+		 * which needs to be handled. We add a check here and return error.
+		 */
+		if (list_length(previousSubPlanList) > 0)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg(
+								"complex joins are only supported when all distributed tables are "
+								"co-located and joined on their distribution columns")));
```

Comment:
wondering whether we should to distinguish this error message somewhat from the usual one to know we're hitting this issue